packages/databyss-services/sources/SourcesProvider.js

```diff
@@ -0,0 +1,72 @@
+import React, { createContext, useContext, useState, useEffect } from 'react'
+import { Text, View } from '@databyss-org/ui/primitives'
+import Loading from '@databyss-org/ui/components/Loading'
+import createReducer from '@databyss-org/services/lib/createReducer'
+import { SourceNotFoundError } from './SourceNotFoundError'
+import { initialState } from './reducer'
+import { fetchSource, saveSource } from './actions'
+
+const useReducer = createReducer()
+
+export const SourceContext = createContext()
+
+const Error = () => (
+  <View height="100%">
+    <View alignSelf="center" justifyContent="center">
+      <Text>No Source Found</Text>
+    </View>
+  </View>
+)
+
+const SourceProvider = ({ children, initialState, reducer }) => {
+  const [state, dispatch] = useReducer(reducer, initialState)
+
+  // provider methods
+  const setSource = source => {
+    // find and update source or
+    // create new cache value and append new source
+    dispatch(saveSource(source))
+  }
+
+  const getSource = id => {
+    if (state.cache[id]) {
+      return state.cache[id]
+    }
+    dispatch(fetchSource(id))
+    return null
+  }
+
+  return (
+    <SourceContext.Provider
+      value={[getSource, setSource, { cache: state.cache }]}
```

Comment:
This was my bad from the strategy. We can simplify to just `[getSource, setSource, state.cache]`